swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+            It will have two dimensions, of structure [epoch/batch, x_step]
+        func_string: A string indicating the type of summary statistic desired.
+            Currently allowable includes 'min', 'max', 'mean', 'median', 'std', and 'range'
+        across: Takes ony 'Indicate whether you'd like to summarise across the bees (as in use case 1 and 2),
+            or summarise across the a value for each point in domain x.
+
+    Returns: a numpy array of two dimensions, of structure [epoch/batch, x_step] that gives the
+        summary evaluation for each point in x space for all the bees at that epoch of training.
+
+    """
+    valid_axes = {'bees', 'x'}
+    if across not in valid_axes:
+        raise ValueError("argument `across` must be one of %r." % valid_axes)
+
+    # Create a simple dict of likely functions
+    funcs = {
+        "mean": np.mean,
+        "std": np.std,
+        "median": np.median,
+        "min": np.min,
+        "max": np.max
+    }
+
+    # Identify the suitable axis to summarise across
+    if across == "bees":
+        axis_choice = 0
+    elif across == "x":
+        axis_choice = len(data.shape) - 1
+
+    # Create summary of data
+    if func_string in funcs.keys():
+        return np.apply_along_axis(funcs[func_string], axis_choice, data)
+    elif func_string == 'range':
+        point_maxs = np.apply_along_axis(np.max, axis_choice, data)
+        point_mins = np.apply_along_axis(np.min, axis_choice, data)
+        return point_maxs - point_mins
+    else:
+        raise ValueError("Function not in currently allowable of min', 'max', 'mean', 'median', 'std', and 'range'")
+
+
+def rms_error(swarm_ypreds: np.ndarray, y: np.ndarray) -> np.ndarray:
+    """
+    This is a function that calculates the root-mean sqared error for a prediction array, given true y.
+    It averages along the all the point predictions in x for a set of predictions.
+    Args:
+        swarm_ypreds: A numpy array of predictions. This can have different numbers of dimensions,
+            eg [bee, epoch, x_step] or [epoch, x_step], so long as the final dimension represents the
+            predictions at different x values, and has same length as y.
+        y: The true values predictions. A 1-d array.
+
+    Returns: A numpy array of dimension one less than the input `swarm_ypreds`, with the evaluated root-mean-squared
+    loss across all predictions.
+    """
+    dims = len(swarm_ypreds.shape)
+    squares = np.square(np.subtract(swarm_ypreds, y))
+    losses = np.apply_along_axis(np.mean, dims-1, squares)
+    return losses
+
+
+def iteration_finder(epoch_scores: [np.ndarray, list],
+                     threshold: float,
+                     threshold_type: str = "nominal",
+                     criterion: str = "first",
+                     direction: str = "below",
+                     reference_iteration: int = 0) -> [int, None]:
```

Comment:
Probably `Optional[int]`

Also, `from typing import Union, Optional` here

Reply:
Great. Didn't know this stuff. Thanks

Reply:
(This is for the Return type btw)